util: detect and warn when using exFAT on MacOS

willcl-ark opened the pull request:
exFAT is known to cause intermittent corruption on MacOS. 

Therefore we should warn when using this fs format for either the blocks or data directories.

See #28552 for more context.

willcl-ark committed:
util: detect and warn when using exFAT on macOS

exFAT is known to cause corruption on macOS. See #28552.

Therefore we should warn when using this fs format for either the blocks
or data directories on macOS.

Co-authored-by: l0rinc <pap.lorinc@gmail.com>

willcl-ark commented:
I did toy with having this trigger a full on `InitError`, but the behaviour is intermittent and seems worst during IBD. Exiting would also likely break things for a fair few users, due to exFAT being used in cross-platform storage devices. Therefore a warning seems more apt.

The root cause seems to be that some calls to `F_FULLFSYNC fcntl` silently failing on exFAT. This appears to stem from the fact that, whilst supported, `F_FULLFSYNC` may not be fully implemented for exFAT:

> F_FULLFSYNC        Does the same thing as fsync(2) then asks the drive to
>                    flush all buffered data to the permanent storage
>                    device (arg is ignored).  This is currently imple-mented implemented
>                    mented on HFS, MS-DOS (FAT), and Universal Disk Format
>                    (UDF) file systems.  The operation may take quite a
>                    while to complete.  Certain FireWire drives have also
>                    been known to ignore the request to flush their
>                    buffered data.

link: https://newosxbook.com/src.jl?tree=xnu&file=/bsd/man/man2/fcntl.2 
and: https://developer.apple.com/library/archive/documentation/System/Conceptual/ManPages_iPhoneOS/man2/fcntl.2.html#//apple_ref/doc/man/2/fcntl

From debugging this seems to manifest from fast processing speed; if I add enough `LogPrintf`'s then it eventually goes away, presumably due to the reduced IBD speed.
This may also indicate that users on slower (CPU) machines may suffer from this less frequently.

We do read and write to the block files from the correct position/offsets at failure but the wrong data is found there. So this appears to be an issue at the kernel level.

I will open a new issue with some outputs from my debugging on this.

willcl-ark commented on src/common/init.cpp:
```diff
@@ -62,6 +63,32 @@ std::optional<ConfigError> InitConfig(ArgsManager& args, SettingsAbortFn setting
 
+        // Warn if we are trying to put the datadir on an exFAT fs on MacOS
+        // This is an upstream issue known to cause bugs, see #28552
+#ifdef __APPLE__
```
I haven't encountered the same error using this drive to sync when formatted with APFS, so I in my opinion, no.

I have come across comments online purporting that certain brands of nvme drives don't play well with apple's (internal) nvme controllers, which can also cause similar issues. But here I feel pretty certain this is an exFAT x MacOS issue, and we may well see the same issues with an internal exFAT drive.

If anyone has a mac with an internal sata/nvme drive which they could format to exFAT and test IBD with, that would be great! (I haven't)

willcl-ark replied:
Oh I should add here, (and probably in #31454) that I tested this same drive using exFAT on Linux, and did not see any corruption to block 450k. By which point every MacOS test had corrupted some data and failed.

edilmedeiros replied:
If my memory is not failing me, I have seem only macOS related issues reporting problems with exFAT external drivers. I tried exFAT on the internal drive of an old 2016 Intel Macbook Pro (original 1GB SSD, APFS on a primary partition and exFAT on a secondary partition for Bitcoin data) and IBD failed on v26. It was about a year ago, I was trying to understand if it was a problem with the external drive I was using at the time or a software problem. Unfortunately, I can't try that again with that computer now.

l0rinc replied:
Even when there isn't any corruption, maybe it would help to warn when [SD cards are used for storage](https://github.com/bitcoin/bitcoin/issues/31882#issuecomment-2661499659)

philmb3487 replied:
Is that really judicious? Plenty of people run Bitcoind on Raspberry Pi's!

l0rinc replied:
Yes, but mostly with SSDs, not SD cards.

l0rinc replied:
Please resolve the comment, seems it's not strictly related to the issue

laanwj replied:
> Oh I should add here, (and probably in https://github.com/bitcoin/bitcoin/issues/31454) that I tested this same drive using exFAT on Linux, and did not see any corruption to block 450k. By which point every MacOS test had corrupted some data and failed.

The bug is specific to macOS and exFAT. Though, exFAT is a trival filesystem meant for things like storing photos (for cameras) and other archival purposes, using it for database-heavy access patterns such as a bitcoin node is a bad idea in general.

marcofleon reviewed (approved):
utACK a7d52d023eee6559d04e2e191f17176f2fe92c00. This happened to me earlier this year during IBD using a mac M3 and an external SSD formatted by default as exFAT. Took me a bit to figure it out, so a warning would've helped.

Reformatting to APFS solved the issue for me. Would that be worth adding to the warning as a suggestion for a solution?

willcl-ark commented:
> Reformatting to APFS solved the issue for me. Would that be worth adding to the warning as a suggestion for a solution?

Yes that could be nice. Another idea that I had suggested was linking to a help page. Perhaps I could dd a section to https://github.com/bitcoin/bitcoin/blob/master/doc/files.md describing the problematic FS/OS combination and add the url to the warning to make things super-easy for users to fix themselves?

cc @fanquake

willcl-ark commented on src/common/init.cpp:
```diff
@@ -62,6 +63,32 @@ std::optional<ConfigError> InitConfig(ArgsManager& args, SettingsAbortFn setting
+            FSType fs_type = GetFilesystemType(check.path);
+            switch(fs_type) {
+                case FSType::EXFAT:
+                    LogWarning("Specified %s \"%s\" is exFAT which is known to have intermittent corruption problems on MacOS.", check.description, fs::PathToString(check.path));
```
Thanks for catching marco. It was indeed a mistake to change this from an `InitError` in my first draft to `LogWarning`.

Will fixup shortly

willcl-ark replied:
Done in 83839a35f07055dd03924b5bdab46bf31df33b35

laanwj commented:
Concept ACK, i think this makes sense in general. exFAT is meant for storing photos, not as a mature filesystem for database workloads.

l0rinc commented:
Concept ACK

marcofleon commented:
Nice, the warning and new addition to the docs lgtm.

ACK ee05e730466f250e131eb92165bcdb8bab68c864

helloscoopa commented:
Concept ACK

hebasto commented:
Concept ACK.

tdb3 commented on src/common/init.cpp:
```diff
@@ -62,6 +64,35 @@ std::optional<ConfigError> InitConfig(ArgsManager& args, SettingsAbortFn setting
+            fs::path path;
+            std::string_view description;
+        };
+        std::array<PathCheck, 2> paths_to_check{{
+            {args.GetDataDirNet(), "data directory"},
+            {args.GetBlocksDirPath(), "blocks directory"}
+        }};
```
Makes sense to cover blocksdir and datadir.
Since the user could also set `-walletdir=<dir>` when starting bitcoind, do you think it would be worth warning for it as well? Much lighter disk usage than blocksdir and datadir, but if we're doubtful of macOS's exFAT implementation, then maybe we would want to warn the user when storing a wallet with it?

edilmedeiros replied:
Agree with this, even though I have never encountered any problems with the wallet files on a Mac with a FAT formatted disk, only on the blocks dir.

tdb3 replied:
Fair point. Either approach (the refactor, or deferring) sounds reasonable. In the big picture, the warnings for blocksdir and datadir add value and I don't think walletdir warning should hold up the PR.

I do wonder how common it actually is that someone (using macOS) is using exFAT for just the walletdir, and would therefore miss the warning. I would imagine most users would be using the default walletdir location, and could therefore see the warning as it pertains to datadir.

tdb3 reviewed (commented):
Concept ACK
Great find!

luke-jr commented on src/common/init.cpp:
```diff
@@ -62,6 +64,35 @@ std::optional<ConfigError> InitConfig(ArgsManager& args, SettingsAbortFn setting
+            FSType fs_type = GetFilesystemType(check.path);
+            switch(fs_type) {
+                case FSType::EXFAT:
+                    InitWarning(strprintf(_("Specified %s \"%s\" is exFAT which is known to have intermittent corruption problems on MacOS. "
```
For GUI, maybe we should check this while they still have the opportunity to change it...

philmb3487 replied:
just one nit, the OS is named 'macOS'

Zero-1729 replied:
I'd be good to add a recommendation here, too, so the user is aware of possible immediate solutions.

```suggestion
                    InitWarning(strprintf(_("Specified %s \"%s\" is exFAT which is known to have intermittent corruption problems on MacOS, use APFS instead. "
```

willcl-ark replied:
> For GUI, maybe we should check this while they still have the opportunity to change it...

To add this check to the *qt/intro.cpp* flow would require duplicating some of the init checks, as well as:

- first creating the chosen directory
- check what FS Type the created dir is
- if EXFAT prompt the user to change filesystems
- either remove the just-created directories (or keep them) depending on response.
- etc.

If we want that kind of logic then I'm happy to add it, but not sure if we need much more than a loud InitWarning myself.

> just one nit, the OS is named 'macOS'

Renamed in all places

gentlee replied:
May be let's not insist so much and use ", consider using APFS instead."

luke-jr reviewed (commented):
Would be ideal if we had a way to reproduce the bug, and can test if the user is actually affected, in case Apple fixes it. But this is better than nothing for now

edilmedeiros commented:
Concept ACK.

Already lost many hours trying to debug this and my conclusion so far is that the FAT driver on macOS is flawed.
Also, it's quite difficult to reproduce this reliably, IBD will fail in a different point every time. Maybe we can detect a broken block in persistent memory and restart IBD from there?

davidgumberg commented:
Concept ACK

I second the suggestion by @tdb3 [above](https://github.com/bitcoin/bitcoin/pull/31453#discussion_r1885462353) to also check `-walletdir`, although I have looked and have not been able to find similar reports of corruption with SQLite and exFAT drives on macOS.

To add a little data pointing in the direction of this bug only existing on macOS, I have synced nodes with datadirs on exFAT drives over both SATA and USB on Fedora 40 and have a long-running bitcoind node that has been running off of an exFAT SATA SSD on Fedora 40 for the past ~6 months and have never experienced any of the symptoms described in https://github.com/bitcoin/bitcoin/issues/28552 or https://github.com/bitcoin/bitcoin/issues/31454.

luke-jr commented on src/util/fs_helpers.cpp:
```diff
@@ -41,6 +41,10 @@
 #endif // WIN32
 
+#ifdef __APPLE__
+#include <sys/mount.h>
```
The manpage says we need `#include <sys/param.h>` too

https://developer.apple.com/library/archive/documentation/System/Conceptual/ManPages_iPhoneOS/man2/statfs.2.html

willcl-ark replied:
added in df1ba101419729aafa382d970ee605e2a6273e26

Zero-1729 commented on doc/files.md:
```diff
@@ -123,6 +125,12 @@ Path           | Description | Repository notes
+
+When choosing a filesystem for the data directory (`datadir`) or blocks directory (`blocksdir`), some filesystems should be avoided:
+
+- **MacOS**: The exFAT filesystem should not be used. There have been multiple reports of database corruption when using exFAT on MacOS for Bitcoin Core. This appears to be due to filesystem-level issues with exFAT on MacOS. See [Issue #31454](https://github.com/bitcoin/bitcoin/issues/31454) for more details.
```
It'd be good to clarify here that it is recommended to use `APFS` instead.

```suggestion
- **MacOS**: The exFAT filesystem should not be used, it is recommended to use the APFS filesystem instead. There have been multiple reports of database corruption when using exFAT on MacOS for Bitcoin Core. This appears to be due to filesystem-level issues with exFAT on MacOS. See [Issue #31454](https://github.com/bitcoin/bitcoin/issues/31454) for more details.
```

willcl-ark replied:
I've avoided explicit recommendation of a FS on purpose here; as far as I know exFAT is the only known-problematic FS on macOS?

Can be convinced otherwise, but leaving as-is for now

Zero-1729 reviewed (commented):
Concept ACK

I encountered this error late last year as well. Thank you for working on this. I think having the warning reported is a good measure.

I've dropped a few comments above.

I tested the patch on an M1 MBP running MacOS v15.2 with an exFAT drive (83839a35f07055dd03924b5bdab46bf31df33b35):

```
$ ./build/src/qt/bitcoin-qt --datadir=/Volumes/Little\ Test/

Warning: Specified data directory "/Volumes/Little Test" is exFAT which is known to have intermittent corruption problems on MacOS. See https://github.com/bitcoin/bitcoin/blob/master/doc/files.md#filesystem-recommendations for more information.
Warning: Specified blocks directory "/Volumes/Little Test/blocks" is exFAT which is known to have intermittent corruption problems on MacOS. See https://github.com/bitcoin/bitcoin/blob/master/doc/files.md#filesystem-recommendations for more information.
2025-01-14 19:45:56.627 bitcoin-qt[55978:5075996] +[IMKClient subclass]: chose IMKClient_Modern
2025-01-14 19:45:56.627 bitcoin-qt[55978:5075996] +[IMKInputSession subclass]: chose IMKInputSession_Modern
...
```

As a minor nit, I do think given this adds a whole bunch of references to `macOS`, it should be formatted as `macOS` instead of `MacOS` for consistency, as `macOS` is used more often across the repo.

Also, I think it'll be helpful to have the disk format warning displayed as a dialog for those running the GUI, same way we get a warning if the disk is too small to accommodate blocks:

<img width="588" alt="Screenshot 2025-01-14 at 19 46 02" src="https://github.com/user-attachments/assets/114812b8-bdf4-4bfb-b213-0d30a00e6f63" />

willcl-ark commented:
df1ba101419729aafa382d970ee605e2a6273e26 squashes the commits together, and also removes the duplicate warning messages, instead printing a single warning with a list of violating directories, e.g. going from:

```
Warning: Specified data directory "/Volumes/sandisk1/signet" is exFAT which is known to have intermittent corruption problems on MacOS. See https://github.com/bitcoin/bitcoin/blob/master/doc/files.md#filesystem-recommendations for more information.
Warning: Specified blocks directory "/Volumes/sandisk1/signet/blocks" is exFAT which is known to have intermittent corruption problems on MacOS. See https://github.com/bitcoin/bitcoin/blob/master/doc/files.md#filesystem-recommendations for more information.
```

to

```
Warning: The following paths are on exFAT which is known to have intermittent corruption problems on MacOS: data directory ("/Volumes/sandisk1/signet"), blocks directory ("/Volumes/sandisk1/signet/blocks"). See https://github.com/bitcoin/bitcoin/blob/master/doc/files.md#filesystem-recommendations for more information.
```


... if both datadir and blocksdir are on exfat on macOS.

tdb3 reviewed (approved):
code review and untested ACK df1ba101419729aafa382d970ee605e2a6273e26

I also think a GUI warning will be beneficial (and would guess is pertinent to a significant number of the users running macOS and exFAT)

While I don't have the hardware to test locally, fwiw cross compilation was successful with Ubuntu 24.04.
```
Cross compiling ....................... TRUE, for Darwin, aarch64
C++ compiler .......................... Clang 18.1.3, /usr/bin/clang++
```

philmb3487 commented:
I don’t think we should bother the user with these gratuitous messages. Has it been demonstrated that SD cards lead to corruption? From what I understand, this issue is about Apple-specific software bugs and it should stick to this.

On Mon, Feb 17, 2025 at 22:32, l0rinc ***@***.***(mailto:On Mon, Feb 17, 2025 at 22:32, l0rinc <<a href=)> wrote:

> @l0rinc commented on this pull request.
>
> ---------------------------------------------------------------
>
> In [src/common/init.cpp](https://github.com/bitcoin/bitcoin/pull/31453#discussion_r1958426704):
>
>> @@ -62,6 +63,32 @@ std::optional<ConfigError> InitConfig(ArgsManager& args, SettingsAbortFn setting
>              fs::create_directories(net_path / "wallets");
>          }
>
> +        // Warn if we are trying to put the datadir on an exFAT fs on MacOS
> +        // This is an upstream issue known to cause bugs, see #28552
> +#ifdef __APPLE__
>
> Yes, but mostly with SSDs, not SD cards.
>
> —
> Reply to this email directly, [view it on GitHub](https://github.com/bitcoin/bitcoin/pull/31453#discussion_r1958426704), or [unsubscribe](https://github.com/notifications/unsubscribe-auth/ABNCTC6GI44QNRHFLNTZNCT2QH6ITAVCNFSM6AAAAABTK2M3GCVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDMMRRGM4DENJYGE).
> You are receiving this because you commented.Message ID: ***@***.***>

philmb3487 commented:
@sipa Thanks for bringing this up. Apple moved the exFat filesystem to userspace and it had nasty side effects. What a terrible idea, what is Apple even thinking?
To be sure, have other drivers been moved to user-space?
For all we know they might also be broken. It needs to be checked, it seems doubtful they would only move one driver over.

On Tue, Feb 18, 2025 at 21:00, l0rinc ***@***.***(mailto:On Tue, Feb 18, 2025 at 21:00, l0rinc <<a href=)> wrote:

> @l0rinc commented on this pull request.
>
> ---------------------------------------------------------------
>
> In [src/common/init.cpp](https://github.com/bitcoin/bitcoin/pull/31453#discussion_r1959806856):
>
>> @@ -62,6 +63,32 @@ std::optional<ConfigError> InitConfig(ArgsManager& args, SettingsAbortFn setting
>              fs::create_directories(net_path / "wallets");
>          }
>
> +        // Warn if we are trying to put the datadir on an exFAT fs on MacOS
> +        // This is an upstream issue known to cause bugs, see #28552
> +#ifdef __APPLE__
>
> Please resolve the comment, seems it's not strictly related to the issue
>
> —
> Reply to this email directly, [view it on GitHub](https://github.com/bitcoin/bitcoin/pull/31453#discussion_r1959806856), or [unsubscribe](https://github.com/notifications/unsubscribe-auth/ABNCTC4IAANWWNUDG5PHKED2QM4JDAVCNFSM6AAAAABTK2M3GCVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDMMRTGY3DSMRUGQ).
> You are receiving this because you were mentioned.Message ID: ***@***.***>

Zero-1729 commented:
I agree with @sipa here, I think the scope and focus of this PR should remain the macOS external SSD exFAT issue which is well documented and easily testable (both the error and @willcl-ark's warning).

Once there is a similar level of replicability for the SD card or similar issue, there can be a separate follow up PR that addresses that.

pablomartin4btc commented on src/common/init.cpp:
```diff
@@ -61,7 +63,48 @@ std::optional<ConfigError> InitConfig(ArgsManager& args, SettingsAbortFn setting
+        }
+
+        if (!exfat_paths.empty()) {
+            InitWarning(strprintf(_("The following paths are on exFAT which is known to have intermittent corruption problems on MacOS: %s. "
```
small nit:
```suggestion
            InitWarning(strprintf(_("The following paths are on exFAT which is known to have intermittent corruption problems on macOS: %s. "
```

willcl-ark replied:
MacOS corrected to macOS in according to #31453 (comment)

pablomartin4btc replied:
Sorry, my bad here, I should've checked properly. That's correct, Apple [rebranded](https://en.wikipedia.org/wiki/macOS) MacOS to macOS in 2016.

pablomartin4btc commented on src/util/fs_helpers.h:
```diff
@@ -14,6 +14,23 @@
+
+/**
+ * Detect filesystem type for a given path.
+ * Currently identifies exFAT filesystems which cause issues on MacOS.
```
same:
```suggestion
 * Currently identifies exFAT filesystems which cause issues on macOS.
```

l0rinc replied:
+1

pablomartin4btc reviewed (commented):
Concept ACK

In case more reviewers consider `-walletdir` and `GUI`, please consider [this](https://github.com/bitcoin/bitcoin/pull/31453#discussion_r1922344525) and [this](https://github.com/bitcoin/bitcoin/pull/31453#discussion_r1922353113) author's comments respectively.

Left a tiny nit, I see was also mentioned by @Zero-1729.

philmb3487 commented:
I don't know. exFAT is the standard filesystem on Android for instance, the world's most widely used OS.

phil / sinetek

Sent with [Proton Mail](https://proton.me/mail/home) secure email.

On Wednesday, April 2nd, 2025 at 6:25 PM, laanwj ***@***.***> wrote:

> @laanwj commented on this pull request.
>
> ---------------------------------------------------------------
>
> In [src/common/init.cpp](https://github.com/bitcoin/bitcoin/pull/31453#discussion_r2024628134):
>
>> @@ -62,6 +63,32 @@ std::optional<ConfigError> InitConfig(ArgsManager& args, SettingsAbortFn setting
>              fs::create_directories(net_path / "wallets");
>          }
>
> +        // Warn if we are trying to put the datadir on an exFAT fs on MacOS
> +        // This is an upstream issue known to cause bugs, see #28552
> +#ifdef __APPLE__
>
>> Oh I should add here, (and probably in [#31454](https://github.com/bitcoin/bitcoin/issues/31454)) that I tested this same drive using exFAT on Linux, and did not see any corruption to block 450k. By which point every MacOS test had corrupted some data and failed.
>
> The bug is specific to macOS and exFAT. Though, exFAT is a trival filesystem meant for things like storing photos (for cameras) and other archival purposes, using it for database-heavy access patterns such as a bitcoin node is a bad idea in general.
>
> —
> Reply to this email directly, [view it on GitHub](https://github.com/bitcoin/bitcoin/pull/31453#discussion_r2024628134), or [unsubscribe](https://github.com/notifications/unsubscribe-auth/ABNCTC2BOGUFFIQOXTKLA2D2XPCJFAVCNFSM6AAAAABTK2M3GCVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDOMZVHEZDKOJVGQ).
> You are receiving this because you were mentioned.Message ID: ***@***.***>

laanwj commented:
> I don't know. exFAT is the standard filesystem on Android for instance, the world's most widely used OS.

No one is preventing you from doing so, but it's still better to use a journaling fs.

achow101 commented:
It looks like this is happening early enough in init that the warning never actually makes it to the GUI, so GUI users (probably a majority of MacOS users) won't actually see the warning.

ismaelsadeeq reviewed (commented):
Concept ACK on this.

This is really helpful for preventing redundant sync attempts pending when the actual issue will be addressed.

---
Sigh, I wouldn’t have synced past 80% three times if this was in.

willcl-ark commented:
@maflcko I'd be happy to split off the doc change if that's preferred (or have someone else take it), but I think that the log warning is actually the most effective part of this PR.

I rebased and moved the checks later in the startup sequence when the GUI is started, so (MacOS) GUI users will see this warning in a big pop-up box like they do with the disk space warnings.

I don't particularly like how this check is so much later in startup; I put it as early as possible before  which is cleaner, but I don't know a good way to pass this message to the GUI after it's started. (I also started with it earlyalso partially because I think we should not support exFAT at all and have this `InitError` rather than just warn).

ismaelsadeeq reviewed (commented):
Code review 690b57560da11a3c25072d4232601bde0cfbc89d


I've tested this on macOS using a custom exFAT filesystem for:

1. `datadir`
2. `-blocksdir`
3. Both `blocksdir` and `datadir`

I think the current implementation is redundant because when the `blocksdir` is just a folder inside `datadir`, we emit two warnings. I think it should be a single warning instead.

I've modified the implementation to achieve this — see:
https://github.com/ismaelsadeeq/bitcoin/commit/ddb02a6beb0ec892c18b43a55d53dc2daf11e555

I've also made `GetFilesystemType` available on all operating systems, not just macOS, as I think it could be useful.

Additionally, I've adjusted the warnings to remove the external link and instead directly include the recommended action. I believe this is clearer and eliminates the need for users to check an external link, which they may likely not do. This is also consistent with the format used for other warnings, such as when the blocks directory size is less than 750 GiB.

@willcl-ark would it be possible to have a functional test for this so that it can be validated in CI i.e we dont have to rely on manual test?
If so, it would be great to include it here( although the test is not a blocker though).

ismaelsadeeq commented on src/init.cpp:
```diff
@@ -1841,6 +1841,50 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
+    }
+
+    if (!exfat_paths.empty()) {
+        InitWarning(strprintf(_("The following paths are on exFAT which is known to have intermittent corruption problems on MacOS: %s. "
```
Not just in this PR  I’ve noticed `InitWarning` emits duplicate warnings when running bitcoind:
one with a timestamp and another without.

When running Qt from the command line: one popup warning and one log entry without a timestamp.

ismaelsadeeq replied:
e.g
```
2025-06-25T06:13:16Z [warning] The following paths are on exFAT which is known to have intermittent corruption problems on MacOS: blocks directory ("/Volumes/StoreJet/blocks_temp/blocks"). Move these directories to a non-exFAT formatted drive to avoid corruption.
Warning: The following paths are on exFAT which is known to have intermittent corruption problems on MacOS: blocks directory ("/Volumes/StoreJet/blocks_temp/blocks"). Move these directories to a non-exFAT formatted drive to avoid corruption.
```

willcl-ark replied:
I'm pretty sure this is actually just two different warning styles both being logged to the same place by the user. I haven't tested, but this code: https://github.com/bitcoin/bitcoin/blob/ad654a4807cd584be9ffcd8640f628ab40cb5170/src/noui.cpp#L33-L46 logs one error to log (with timestamp) and when running the daemon the same error (without timestamp) to `stderr`.

When running daemon in the foreground, both messages appear printed in the console. But IMO this makes sense for the daemonised process to outpue a message like this not only to the logfile, but also stderr (to be read/processed by other tools).

ismaelsadeeq replied:
Yes

willcl-ark commented:
Thanks for testing.

> I think the current implementation is redundant because when the `blocksdir` is just a folder inside `datadir`, we emit two warnings. I think it should be a single warning instead.

Could you provide an example of the two warnings? It should print one warning listing the two paths.

> @willcl-ark would it be possible to have a functional test for this so that it can be validated in CI i.e we dont have to rely on manual test? If so, it would be great to include it here( although the test is not a blocker though).

I suppose it might be possible to add, but having a test which repartitions a drive, and creates a new exFAT volume feels icky to me! I'm happy with having this warning documented and a best-effort check for it on startup.

willcl-ark commented:
> No need to mention the warning for the blocksdir since it just a folder in the datadir; It's whats being checked in [ismaelsadeeq@ddb02a6](https://github.com/ismaelsadeeq/bitcoin/commit/ddb02a6beb0ec892c18b43a55d53dc2daf11e555)

Ok, we now check if blocksdir is inside datadir, and removed the long URL from the warning.

ismaelsadeeq reviewed (commented):
Thanks for taking my suggestion @willcl-ark

Your modified string is even better!

Just a couple of comments and then I will re-test again.

ismaelsadeeq commented on src/init.cpp:
```diff
@@ -1841,6 +1841,55 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
+        }
+    }
+
+    if (!exfat_paths.empty()) {
+        InitWarning(strprintf(_("The following paths are on exFAT which is known to have intermittent corruption problems on MacOS: %s. "
+            "Move these directories to a non-exFAT formatted drive to avoid corruption."),
+            util::Join(exfat_paths, ", ")));
+    }
```
nitty-nit: right now the ouput seems to be plural, i.e directories when it's one so this might be better?
```suggestion
if (!exfat_paths.empty()) {Add commentMore actions
        for (const auto& path : exfat_paths) {
            InitWarning(strprintf(_("The specified %s is on exFAT which is known to have intermittent corruption problems on macOS. "
                                    "Restart with non-exFAT formatted drive to avoid corruption."),
                                    path));
        }
    }
```

willcl-ark replied:
Going to leave this nit, as I don't think it's worth exatra logic here

ismaelsadeeq reviewed (commented):
Code review and tested ACK 2ad18fc9784e6a37e453bf844ba10c4e46f54754

bitcoind
```
2025-06-25T06:12:15Z [warning] The following paths are on exFAT which is known to have intermittent corruption problems on MacOS: data directory ("/Volumes/StoreJet/bitcoin_temp"). Move these directories to a non-exFAT formatted drive to avoid corruption.
```

gui 
<img width="577" alt="Screenshot 2025-06-25 at 07 29 37" src="https://github.com/user-attachments/assets/1e1340e6-9e3e-4aed-beb6-8ab5a5eed678" />

```

luke-jr commented on src/init.cpp:
```diff
@@ -1841,6 +1841,55 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
+    fs::path blocks_dir = args.GetBlocksDirPath();
+    std::vector<PathCheck> paths{{data_dir, "data directory"}};
+    // Don't mention blocks dir if it's inside datadir
+    if (fs::relative(blocks_dir, data_dir / "blocks").string() != ".") {
```
What if it's datadir/blocks-alternative-name or something?

luke-jr commented on src/init.cpp:
```diff
@@ -1841,6 +1841,55 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
+    fs::path data_dir = args.GetDataDirNet();
+    fs::path blocks_dir = args.GetBlocksDirPath();
+    std::vector<PathCheck> paths{{data_dir, "data directory"}};
+    // Don't mention blocks dir if it's inside datadir
```
Does macOS not allow mounting subdirectories? (Maybe a bind mount) eg, maybe datadir isn't exFAT, but blocks is?

willcl-ark replied:
Agree. I removed this "optimisation" in eb68bb48ab2bb163f20dc7430d0401d88ca43f4a

We now just check both directories to avoid missing any as you describe (and if they are renamed as per https://github.com/bitcoin/bitcoin/pull/31453#discussion_r2171706368 )

ismaelsadeeq commented:
It will be nice if this get in to v30, cc all previous reviewers @maflcko perhaps add the v30.0 milestone

ismaelsadeeq reviewed (approved):
reACK eb68bb48ab2bb163f20dc7430d0401d88ca43f4a

Also retested on both `bitcoin-qt` and `bitcoind` binaries release build and I get the same warnings when datadir/blocksdir or both are on exFat drive format

l0rinc commented on src/init.cpp:
```diff
@@ -1866,6 +1866,35 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
+    check_fs(args.GetBlocksDirPath(), "blocks directory");
+
+    if (!exfat_paths.empty()) {
+        InitWarning(strprintf(_("The following paths are on exFAT which is known to have intermittent corruption problems on macOS: %s. "
```
nit: "are on exFAT" sounds weird to me + missing comma:
```suggestion
        InitWarning(strprintf(_("The following paths are using exFAT, which is known to have intermittent corruption problems on macOS: %s. "
```

l0rinc commented on src/util/fs_helpers.cpp:
```diff
@@ -298,3 +303,17 @@ std::optional<fs::perms> InterpretPermString(const std::string& s)
 }
+
+#ifdef __APPLE__
+FSType GetFilesystemType(const fs::path& path) {
```
nit: non-standard formatting

l0rinc reviewed (changes requested):
Seeing this planned for [V30](https://github.com/bitcoin/bitcoin/milestone/72) I reviewed it a bit more thoroughly.
My Concept ACK remains, but I think we can document and simplify and modernize it a bit more.

<details>
<summary>Suggestions patch</summary>

```patch
diff --git a/doc/files.md b/doc/files.md
index 21a73d7757..9bdb660b57 100644
--- a/doc/files.md
+++ b/doc/files.md
@@ -164,7 +164,5 @@ This table describes the files installed by Bitcoin Core across different platfo
 - On Windows, binaries have a `.exe` suffix (e.g., `bitcoin-cli.exe`).
 
 ## Filesystem recommendations
-
-When choosing a filesystem for the data directory (`datadir`) or blocks directory (`blocksdir`), some filesystems should be avoided:
-
-- **macOS**: The exFAT filesystem should not be used. There have been multiple reports of database and other corruption when using exFAT on macOS for Bitcoin Core. This appears to be due to filesystem-level issues with exFAT on the macOS operating system. See [Issue #31454](https://github.com/bitcoin/bitcoin/issues/31454) for more details.
+When choosing a filesystem for the data directory (`datadir`) or blocks directory (`blocksdir`) on **macOS**,the `exFAT` filesystem should be avoided.
+There have been multiple reports of database corruption and data loss when using this filesystem with Bitcoin Core, see [Issue #31454](https://github.com/bitcoin/bitcoin/issues/31454) for more details.
diff --git a/src/init.cpp b/src/init.cpp
index e136a32b49..2c2e3b77d5 100644
--- a/src/init.cpp
+++ b/src/init.cpp
@@ -1867,32 +1867,23 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
     }
 
 #ifdef __APPLE__
-    // Warn about exFAT filesystem usage on macOS
-    std::vector<std::string> exfat_paths;
-    std::vector<std::string> error_paths;
-
-    auto check_fs = [&](const fs::path& path, std::string_view desc) {
-        FSType fs_type = GetFilesystemType(path);
-        std::string path_desc = strprintf("%s (\"%s\")", desc, fs::PathToString(path));
-        if (fs_type == FSType::EXFAT) {
-            exfat_paths.push_back(path_desc);
-        } else if (fs_type == FSType::ERROR) {
-            error_paths.push_back(path_desc);
-        }
-    };
-
-    check_fs(args.GetDataDirNet(), "data directory");
-    check_fs(args.GetBlocksDirPath(), "blocks directory");
-
-    if (!exfat_paths.empty()) {
-        InitWarning(strprintf(_("The following paths are on exFAT which is known to have intermittent corruption problems on macOS: %s. "
-            "Move these directories to a non-exFAT formatted drive to avoid corruption."),
-            util::Join(exfat_paths, ", ")));
-    }
-
-    if (!error_paths.empty()) {
-        LogInfo("Failed to detect filesystem type for: %s", util::Join(error_paths, ", "));
-    }
+    auto check_and_warn_fs{[&](const fs::path& path, std::string_view desc) {
+        const auto path_desc{strprintf("%s (\"%s\")", desc, fs::PathToString(path))};
+        switch (GetFilesystemType(path)) {
+        case FSType::EXFAT:
+            InitWarning(strprintf(_("The %s path uses exFAT, which is known to have intermittent corruption problems on macOS. "
+                "Move this directory to a different filesystem to avoid data loss."), path_desc));
+            break;
+        case FSType::ERROR:
+            LogInfo("Failed to detect filesystem type for %s", path_desc);
+            break;
+        case FSType::OTHER:
+            break;
+        }
+    }};
+
+    check_and_warn_fs(args.GetDataDirNet(), "data directory");
+    check_and_warn_fs(args.GetBlocksDirPath(), "blocks directory");
 #endif
 
 #if HAVE_SYSTEM
diff --git a/src/util/fs_helpers.cpp b/src/util/fs_helpers.cpp
index ef7cc78896..b5dc1cb332 100644
--- a/src/util/fs_helpers.cpp
+++ b/src/util/fs_helpers.cpp
@@ -305,13 +305,11 @@ std::optional<fs::perms> InterpretPermString(const std::string& s)
 }
 
 #ifdef __APPLE__
-FSType GetFilesystemType(const fs::path& path) {
-    struct statfs fs_info;
-    if (statfs(path.c_str(), &fs_info) != 0) {
+FSType GetFilesystemType(const fs::path& path)
+{
+    if (struct statfs fs_info; statfs(path.c_str(), &fs_info)) {
         return FSType::ERROR;
-    }
-
-    if (strcmp(fs_info.f_fstypename, "exfat") == 0) {
+    } else if (std::string_view{fs_info.f_fstypename} == "exfat") {
         return FSType::EXFAT;
     }
     return FSType::OTHER;
diff --git a/src/util/fs_helpers.h b/src/util/fs_helpers.h
index 7423da82a4..4d11f82809 100644
--- a/src/util/fs_helpers.h
+++ b/src/util/fs_helpers.h
@@ -23,7 +23,7 @@ enum class FSType {
 
 /**
  * Detect filesystem type for a given path.
- * Currently identifies exFAT filesystems which cause issues on MacOS.
+ * Currently identifies exFAT filesystems which cause issues on macOS.
  *
  * @param[in] path The directory path to check
  * @return FSType enum indicating the filesystem type

```

</details>

willcl-ark commented on src/util/fs_helpers.cpp:
```diff
@@ -298,3 +303,17 @@ std::optional<fs::perms> InterpretPermString(const std::string& s)
+        return FSType::ERROR;
+    }
+
+    if (strcmp(fs_info.f_fstypename, "exfat") == 0) {
```
Yeah, I also could not find such documentation when I first wrote the patch, and ended up making a simple c program to just call `getfsstat` on my mounted exfat disk, to get the info out to use here. I am not at home now though so can't even pass said program to you right now.

> I'm wondering if it's always written like this or if it's sometimes exFAT (like you refer to it) or [ExFAT](https://support.apple.com/guide/disk-utility/file-system-formats-dsku19ed921c/mac) (like the apple doc does sometimes) - or even if it would/should work with something like [relan/exfat](https://github.com/relan/exfat?rgh-link-date=2025-08-07T23%3A57%3A47Z) (found this accidentally, ignore if not important)

It's a fair question, I don't know how brittle this might be, but can't imagine that it would change much?

l0rinc commented on doc/files.md:
```diff
@@ -160,3 +162,7 @@ This table describes the files installed by Bitcoin Core across different platfo
 - README and bitcoin.conf files are included in binary releases but not installed in source builds.
 - On Windows, binaries have a `.exe` suffix (e.g., `bitcoin-cli.exe`).
+
+## Filesystem recommendations
+When choosing a filesystem for the data directory (`datadir`) or blocks directory (`blocksdir`) on **macOS**,the `exFAT` filesystem should be avoided.
```
nit, if you touch again:
```suggestion
## Filesystem recommendations

When choosing a filesystem for the data directory (`datadir`) or blocks directory (`blocksdir`) on **macOS**, the `exFAT` filesystem should be avoided.
```
(my mistake)

willcl-ark replied:
Done in db3228042b2278faf7713a6ed40cdbf8e62b8fd4

l0rinc commented on src/util/fs_helpers.cpp:
```diff
@@ -298,3 +303,15 @@ std::optional<fs::perms> InterpretPermString(const std::string& s)
+{
+    if (struct statfs fs_info; statfs(path.c_str(), &fs_info)) {
+        return FSType::ERROR;
+    } else if (std::string_view{fs_info.f_fstypename} == "exfat") {
```
Few references I found that confirm this is the correct value:
* https://github.com/KDE/kcoreaddons/blob/master/src/lib/io/kfilesystemtype.cpp#L38
* https://docs.getlogga.com/building-blocks/access-logs#example-mount-payloads
* https://github.com/rfjakob/gocryptfs/blob/master/internal/syscallcompat/quirks_darwin.go#L12

l0rinc commented on src/util/fs_helpers.cpp:
```diff
@@ -298,3 +303,15 @@ std::optional<fs::perms> InterpretPermString(const std::string& s)
 }
+
+#ifdef __APPLE__
+FSType GetFilesystemType(const fs::path& path)
```
nit: this may still be a bit too general for the current problem, I'm fine with it as it is, but if you touch again consider:
```C++
bool IsExfatFilesystem(const std::filesystem::path& p)
```
Which would simplify the PR even more.

l0rinc reviewed (approved):
Modified:
```C++
FSType GetFilesystemType(const fs::path& path)
{
    return FSType::EXFAT;
}
```
it correctly logs:
```
Warning: The data directory ("/Users/lorinc/bitcoin/demo") path uses exFAT, which is known to have intermittent corruption problems on macOS. Move this directory to a different filesystem to avoid data loss.
Warning: The blocks directory ("/Users/lorinc/bitcoin/demo/blocks") path uses exFAT, which is known to have intermittent corruption problems on macOS. Move this directory to a different filesystem to avoid data loss.
```
(note the lack of timestamp)

and for:
```C++
FSType GetFilesystemType(const fs::path& path)
{
    return FSType::ERROR;
}
```
it logs:
```
2025-08-08T17:41:12Z Failed to detect filesystem type for data directory ("/Users/lorinc/bitcoin/demo")
2025-08-08T17:41:12Z Failed to detect filesystem type for blocks directory ("/Users/lorinc/bitcoin/demo/blocks")
```

and for the original code or:
```C++
FSType GetFilesystemType(const fs::path& path)
{
    return FSType::OTHER;
}
```

I'm not getting any additional warnings.

ACK 504fe8a3e0a3def67a1ed3e309992d0115cf13cf

l0rinc commented:
ACK db3228042b2278faf7713a6ed40cdbf8e62b8fd4

achow101 commented on src/init.cpp:
```diff
@@ -1866,6 +1866,26 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
+    }};
+
+    check_and_warn_fs(args.GetDataDirNet(), "data directory");
+    check_and_warn_fs(args.GetBlocksDirPath(), "blocks directory");
```
This now causes two warning dialogs to appear consecutively when the blocksdir is a subdirectory of the datadir.

ismaelsadeeq reviewed (commented):
reACK db3228042b2278faf7713a6ed40cdbf8e62b8fd4

Could be nice if https://github.com/bitcoin/bitcoin/pull/31453#discussion_r2264040626 will be fixed as well, but not a blocker for me.

fanquake commented:
Translations followup in #33193.
